File: mesmaeker-java-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/ValidatingOngoingClause.java
package fr.xgouchet.mesmaeker.mockito;

import java.util.function.Function;

import static fr.xgouchet.mesmaeker.core.ValidatingUtils.verifyFunctionThrows;
import static org.assertj.core.api.Java6Assertions.assertThat;

/**
 * @author Xavier F. Gouchet
 */
public class ValidatingOngoingClause<T, O> implements OngoingClause<T, O> {

    private final Function<T, O> function;
    private final T subject;

    /**
     * @param instance
     * @param function
     */
    ValidatingOngoingClause(T instance, Function<T, O> function) {
        this.subject = instance;
        this.function = function;
    }

    @Override
    public OngoingClause<T, O> thenReturn(O value) {
        assertThat(function.apply(subject)).isEqualTo(value);
        return this;
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    @SuppressWarnings({"unchecked", "varargs"}) @Override
    public OngoingClause<T, O> thenReturn(O initialValue, O... values) {
        assertThat(function.apply(subject)).isEqualTo(initialValue);

        for (O value : values) {
            assertThat(function.apply(subject)).isEqualTo(value);
        }
        return this;
    }

    @Override
    public OngoingClause<T, O> thenThrow(Throwable... throwables) {
        for (Throwable t : throwables) {
            verifyFunctionThrows(subject, function, t);
        }
        return this;
    }

    @Override
    public OngoingClause<T, O> thenThrow(Class<? extends Throwable> throwableType) {
        verifyFunctionThrows(subject, function, throwableType);
        return this;
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    @SuppressWarnings({"unchecked", "varargs"}) @Override
    public OngoingClause<T, O> thenThrow(Class<? extends Throwable> toBeThrown, Class<? extends Throwable>... nextToBeThrown) {
        verifyFunctionThrows(subject, function, toBeThrown);

        for (Class<? extends Throwable> t : nextToBeThrown) {
            verifyFunctionThrows(subject, function, t);
        }
        return this;
    }
}

File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/ValidatingOngoingClause.kt
package fr.xgouchet.mesmaeker.mockito

import java.util.function.Function

import fr.xgouchet.mesmaeker.core.ValidatingUtils.verifyFunctionThrows

/**
 * @author Xavier F. Gouchet
 */
class ValidatingOngoingClause<T, O>
/**
 * @param instance
 * @param function
 */
internal constructor(private val subject: T,
                     private val function: (T) -> O) : OngoingClause<T, O> {

    override fun thenReturn(value: O): OngoingClause<T, O> {
        val actual = function(subject)
        require(actual == value, { "Oups" })
        return this
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    override fun thenReturn(value: O,
                            vararg values: O)
            : OngoingClause<T, O> {
        var actual = function(subject)
        require(actual == value, { "Oups" })

        for (nextValue in values) {
            actual = function(subject)
            require(actual == nextValue, { "Oups" })
        }
        return this
    }

    override fun thenThrow(vararg throwables: Throwable)
            : OngoingClause<T, O> {
        for (t in throwables) {
            verifyFunctionThrows(subject, function, t)
        }
        return this
    }

    override fun thenThrow(throwableType: Class<out Throwable>)
            : OngoingClause<T, O> {
        verifyFunctionThrows(subject, function, throwableType)
        return this
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    override fun thenThrow(toBeThrown: Class<out Throwable>,
                           vararg nextToBeThrown: Class<out Throwable>)
            : OngoingClause<T, O> {
        verifyFunctionThrows(subject, function, toBeThrown)

        for (t in nextToBeThrown) {
            verifyFunctionThrows(subject, function, t)
        }
        return this
    }
}
